### DISCUSSION TITLE
Add role mapping abillity to oauth SSO feature

### DISCUSSION THREAD

**Discussion by Cyb4Black** [open-webui/open-webui#5897] [Ideas]
We wanted to manage roles from our internal Keycloak instance, so we can block most users in the realm from Open WebUI and only allow certain people (the people working on ai, who could've thought...) to use it.
Therefore I forked your project and implemented two new ENV Variables ENABLE_OAUTH_ROLE_MAPPING and OAUTH_ROLES_CLAIM.
I also added handling them within the oauth callback function already including the handling of nested claims (as Keycloak puts the roles of a client 3 levels deep into nested claim).
Everything works perfectly fine.

Question is actually: Do you want me throw this oauth extension in as a PR? I'd love to create one and also add some lines to the docs repo.

**Comment by Cyb4Black** [comment]
Maybe one thing to discuss:
I implemented it with direct mapping, so oauth provider needs to deliver roles pending/user/admin.
I could switch this to an implementation like in jupyter hub.
That would be adding 2 ENVs that can hold comma seperated lists of:
- allowed roles -> oauth roles that map to the internal role "user"
- admin roles -> oauth roles that will be mapped to the internal role "admin"

Dont care about "pending". Its simply the fallback for having no oauth role, no need to set it manually.

**Comment by justinh-rahb** [comment]
Long as its thoroughly tested and won't break existing OAuth setups, yes we'd love a PR for this feature!

**Comment by nickolay** [comment]
Thanks for contributing this! Could you share your keycloak setup as well?

I think in default configuration Keycloak returns roles in `realm_access.roles` of the access token https://github.com/keycloak/keycloak/issues/14617 , but `OAUTH_ROLES_CLAIM=realm_access.roles` doesn't seem to have effect.

After reading https://github.com/open-webui/open-webui/blob/21b8ca345904005f4b87666e4e0ac5bb8df309ad/backend/open_webui/utils/oauth.py#L137-L141 and https://docs.authlib.org/en/latest/client/starlette.html#starlette-openid-connect I guess I have to use "Add to ID token" in keycloak, even though the [docs](https://docs.openwebui.com/tutorials/features/sso/#oauth-role-management) say the provider must be "configured to return roles in the *access token*"

**Reply by Cyb4Black** [reply]
Actually the roles are not mapped to ~`realm_access.roles`, but `realm_access.<client_id>.roles`~ in keycloak. So this maybe already solving the issue.

I have it mapped to both access and and ID token, I will evaluate implementation and docs asap to check if I got something mixed up.

Edit:
nvm... I mixed up realm_access and resource_access in my head while writing above...
Let me get some sleep and I'll revisit this...

**Reply by nickolay** [reply]
My experiments with [authlib's demo](https://github.com/authlib/demo-oauth-client/blob/master/starlette-google-login/app.py) confirm that [authlib's `authorize_access_token`](https://github.com/lepture/authlib/blob/64c3eed442a86a7ff6bdc8f63e119cf201213eed/authlib/integrations/starlette_client/apps.py#L81-L85) gives you the raw data via `token["access_token"]` and parsed id_token in `token['userinfo']`.

So until I configured keycloak via _Client -> Client Scopes -> [clientid]-dedicated -> Add mapper (from predefined mappers) -> "realm roles" -> enable "Add to ID token"_ the realm roles (`"realm_access": {"roles": [...]}`) were available in `token["access_token"]` (decoded manually to confirm), but not in `token['userinfo']`. (When explicitly requested via `'scope': 'openid email profile roles'`.)

So I think the docs should say the provider must return the roles in the ID token, not access token?

**Reply by nickolay** [reply]
While I still believe my findings above to be valid, the more orthodox way of setting up roles in Keycloak seems to be:
1) create a client-specific role named "user" (to avoid having to configure [`OAUTH_ALLOWED_ROLES`](https://docs.openwebui.com/tutorials/features/sso/#oauth-role-management) in open-webui)
2) assign it to a keycloak group as needed (by going to the group and using *Assign role* with *Filter by client roles*, NOT by associating the client role with an existing realm role, as [it works the other way around](https://www.keycloak.org/docs/26.0.5/server_admin/#_composite-roles)),
3) set "Add to ID token" either in the global `client roles` mapper, or create one for your client: Client -> Client Scopes -> [clientid]-dedicated -> Add mapper (from predefined mappers)
4) use `OAUTH_ROLES_CLAIM=resource_access.YOUR_OAUTH_CLIENT_ID.roles`

I'll test it early next week.

**Reply by Cyb4Black** [reply]
Your findings are correct, docs should say id token, not access token, though we could parametrize and let operator choose which one to use.

Your "orthodox" usage descriotion is exactly what i did.
Have client openwebui, have groups owa-users and owa-admins, have associated client roles to the groups. And have global client role mapper set to include roles in id token (and userinfo endpoint)

**Comment by tkg61** [comment]
Is this feature now more possible due to 0.4.0 being out? Groups are great but mapping to SSO groups would be a game changer.

**Reply by Cyb4Black** [reply]
it's already implemented and released for a couple of weeks.

**Reply by qdrop17** [reply]
Is there a reference implementation or configuration example available? I need clarification on how OpenWebUI groups map to OIDC claims, particularly for Azure AD integration.

Our goal: Map AzureAD groups to OpenWebUI groups.

I'd create a PR for the docs as soon as I understand the framework.

**Reply by tkg61** [reply]
I think I misunderstood this ticket. I looked through the Auth code above and I see where the env for admin vs. user roles are called out (definitely didn’t know about it so docs would be great), but I wonder if we need a different issue to now take sso group membership and map to existing openwebui groups…

**Reply by tkg61** [reply]
i have made a new feature request for this #7102
